Reorganize some of the BehaviorTree-related code

mjeronimo opened the pull request:
## Description
* Eliminate the NavigateToPoseBehaviorTree class, merging it with the BehaviorTreeEngine
* Put clear_entirely service client with the others in nav2_util
* Move RecoveryNode to nav2_behavior_tree with the other BT nodes

## Future work that may be required in bullet points

* Port the 3.0 version of the Behavior Tree library (which supports plugins)
* Implement plug-ins

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/behavior_tree_engine.hpp:
```diff
@@ -69,6 +70,18 @@ class BehaviorTreeEngine
 
 protected:
+  // Methods used to register as (simple action) BT nodes
+  BT::NodeStatus globalLocalizationServiceRequest();
```
`globalLocalizationServiceRequest` and `initialPoseReceived` arent used, that was part of my proposal to just delete them. They break important Navigation encapsulation concepts.

mjeronimo replied:
Ok, if they aren't used, I can delete them.

SteveMacenski replied:
The navigator shouldn't be tightly coupled with the positioning system, hooks are OK but the core navigator code *in my opinion* shouldn't contain positioning specific items. Since this is the "core" of the BT (and in future loading of nodes) this seems like exactly the place this shouldn't be. Probably as a utility? 

Yes the factory items are all good since we haven't gotten to plugin-izing things.

mjeronimo replied:
How about the localization-related nodes as a separate plugin from the navigation-related nodes? For now, both would be registered in the BehaviorTreeEngine, but soon, they would be separated out so that the BehaviorTreeEngine is generic and each plug-in loads its own node types.

SteveMacenski replied:
I don't think localization-based nodes should exist except for checking-state to trigger something like "stop navigator, I'm delocalized an entering a localization-driven motion mode to re-establish localization". The navigator's job is to take positioning and go to a place according to some rule set that the BT defines. I think that breaks the encapsulation of a pure-navigator. 

It seems to me it would make most sense if the autolocalization mode was an action server that could be called from a node that checks localization quality and triggers, but it itself shouldn't really be in the Navigators behavior tree

mjeronimo replied:
Ok, I understand your point about trying to keep the navigation BT "pure." However, if we made the autolocalization an action server and chose to implement this action server with a BT (similar to how the navigation is implemented using a BT), we'd still need to define the localization-related BT nodes (presumably using a plugin-model) for the auto-localization action server. We've then also introduced yet another node. How about instead using the XML include mechanism supported in the 3.0 library to move the auto-localization sub-tree into its own XML file. The net effect is that the navigation BT references only a single, external node/subtree. We have the benefit of encapsulation without the overhead of introducing another node and action server.

SteveMacenski replied:
Why would autolocalization need to be a behavior tree? The package is called `nav2_bt_navigator` and that has a set of expectations on it - trying to have a package that does everything is probably not a good way to go. 

Auto-localization is using experimental tech that only works for 1 robot model in simulation (safely), I dont think that is mature enough to justify making compromises for in design. If there's some hacky-way we get auto-localization into the stack, I think that's probably more appropriate since its not something most users will be using at this point as it won't work on most people's configurations.

SteveMacenski replied:
Reading back over this: the 2 fns in question both just call services or topics. Why do they have to be in the navigator? The autolocalization module should be more than capable of creating it own service clients, correct @mjeronimo

mjeronimo commented:
@SteveMacenski Sorry about the duplicated effort. I had some code from my other PR that I split off. 

Going forward, I created a meta-ticked: https://github.com/ros-planning/navigation2/issues/958 for the BT work. Can you add the items that you think we should do. Then, we can assign names and split up the work. Thanks.

mjeronimo commented:
@SteveMacenski How about we proceed with this PR and then you take care of the clear costmap change that you mention as a follow-up?

SteveMacenski commented:
No worries, it happens. 

Clearing costmap recovery as another PR is fine with me. Only thing to me that is worth blocking for discussion is removing those extra functions in the thread above

mjeronimo commented:
If OK with @mhpanah, I can remove those functions and then we can introduce later when we're actually using them. Also, if they are localization-specific, they'd be contained in a separate plug-in from the navigation-specific plugin.

mhpanah commented:
@mjeronimo @SteveMacenski Very soon we will re-introduce the auto-localization functionality back to the stack. My vote is to keep it, but if we prefer to remove it now and reintroduce it later then I think that should happen in a separate PR.

SteveMacenski commented:
Ok,
I'm worried about having the autolocalization in the default stack configs since its using a highly experimental technology under the hood that hasn't been widely trained against real world environments, and is uniquely trained on the turtlebot3 model that doesn't generalize across other hardware platforms - whereas Navigation is meant to be generalized. Having it in the Navigator seems dangerous. 

If its a research instrument, then I think it should be plugged into the Navigator via other means than some highly active support. 

But if I'm alone in that, maybe ignore me and do it anyhow. I believe that the defaults and support needs to be for stable and mature technologies people can rely on -- and then also have some experimental packages with instructions about how they should intentionally and deliberately install/configure for their application. That barrier to entry I think is important so someone doesn't do something stupid and run a jackal off their office building because it didnt see the glass

mjeronimo commented:
@SteveMacenski @mhpanah How about an offline discussion to resolve? I think it would be higher bandwidth and easier to explain in a call.

SteveMacenski commented:
Sure, but we need to make sure we have a paper trail for these types of conversations in tickets for posterity.

mjeronimo commented:
Yes, I agree. Let's talk and then summarize here.

SteveMacenski commented:
I'll wait until the autolocalization stuff is done and I can see then what I might propose to make the bt navigator cleaner. I don't want to block this since some clear costmap recovery work I'm going to do next week needs this merged
